RFC: Authorization support (#133)
modelcontextprotocol/specification: docs/specification/draft/basic/authentication.md

---
title: Authentication
type: docs
weight: 15
---

{{< callout type="info" >}}
**Protocol Revision**: {{< param protocolRevision >}}
{{< /callout >}}

## Overview

The Model Context Protocol currently supports authentication through OAuth 2.0. Authentication is only available when using HTTP transport. Authentication occurs at the transport level. This specification defines how clients authenticate with MCP servers using standard OAuth 2.0 flows.

## Authentication Flow

> MCP servers **MAY** require authentication for any request. When authentication is required, servers **MUST** respond with *HTTP 401 Unauthorized*. Clients **MUST** then initiate the OAuth 2.0 authentication flow.

Reviewer: ```suggestion
MCP servers **MAY** require authentication for any request. When authentication is required and has not been provided, servers **MUST** respond with *HTTP 401 Unauthorized*. Clients **MUST** then initiate the OAuth 2.0 authentication flow.
```

The basic authentication flow is as follows:

```mermaid
sequenceDiagram
    participant C as Client
    participant M as MCP Server
    participant A as OAuth Server

    C->>M: Any MCP Request
    M->>C: HTTP 401 Unauthorized

>     C->>M: GET /authorize?redirect_uri=http://localhost:1234/callback

Reviewer 2: Who decides `http://localhost:1234/callback` in this flow? Did the 401 return this or is the client making up the servers callback url?

Reviewer 3: I believe this is client making up it's own callback URL so when the OAuth process is complete it can retrieve the session token (correct me if I am wrong @dsp-ant)

Reviewer 2: Got it, i see `Redirect to validated redirect_uri with session token` at the bottom.

    Note over M: Validate redirect_uri

>     M->>C: Redirect to OAuth Server

Reviewer: The details about redirection are a bit scant here… does this mean literally HTTP 30x?

And should the `GET /authorize` be happening in a web browser context?

Author: Good catch, I'l add more comments. GET /authorize happens either in a browser or outside of a browser, up to client. The redirected URl needs to happen in the browser. I think clients will do /authorize in the browser already and rely on redirect mechanics.

>     Note over M: Validate redirect_uri

Reviewer 4: To double-check my understanding: does the client provide the `redirect_uri`? (I.e., a local client might host a temporary server available at `localhost:NNNN` to receive the callback, whereas a client running on a HTTP server might provide `https://my-mcp-host.com/`?)

I ask because we do (something similar to) a OAuth2.0 device flow on mcp.run that _might_ sidestep the need to have the callback land back at a local server. (The MCP server requests a code/authorization url from mcp.run, then polls for the code to become valid while presenting the authorization url to the user.) I'm not sure if it's in-scope for this proposal, though!

Reviewer 5: Yes, exactly.  Often times "public" desktop apps embed a web server, bind to localhost, and use the resulting port for their callback URL.  Server-side apps would just use their domain and standard port.

One benefit of decoupling the MCP Server and OAuth server, is that authorization flows (and introducing new ones, such as the device flow) don't impact the MCP server itself.   That functionality is delegated to the OAuth server, which can be extended as necessary to support new forms of authentication, new flows, etc - without impacting the MCP server itself (just as is the case with typical HTTP/REST APIs).

I suspect the reason to show the redirect-based authorization code flow is simply because they are more common.  We should avoid making assumptions that limit interactions to just that flow, however.

    C->>A: OAuth Authorization Request
    A->>M: Authorization Code (via redirect)
    M->>A: Exchange Code for Tokens
    A->>M: Access Token + Refresh Token
    M->>C: Redirect to validated redirect_uri with session token

>     C->>M: Subsequent requests with session token

Reviewer 4: Possibly something to consider here, depending on whether keeping the SSE transport browser-compatible is a goal: `EventSource` doesn't support setting additional headers, at least in the browser. (It _does_ support sending `Cookie` headers automatically using the `{ withCredentials }` `EventSourceInit` option, however.)

Reviewer: This is a really good point. 😕 

I don't love cookies as a replacement here. Possibly the most correct answer is, "just use WebSocket," but of course that has its own problems…

Reviewer: Maybe the use of a package like [eventsource](https://www.npmjs.com/package/eventsource) is an acceptable workaround.

Reviewer: Used `eventsource` in this PR: https://github.com/modelcontextprotocol/typescript-sdk/pull/134

    Note over C,M: Begin standard MCP message exchange
```

### Authorization Endpoints

MCP servers **MUST** expose the following OAuth-related endpoints:

1. `/authorize` - Initial authorization endpoint

> 2. `/token` - Token exchange endpoint

Reviewer 6: Where would the `<mcp_server>/token` endpoint be used?

> 2. `/token` - Token exchange endpoint

Author: Yes, it would be easier. The only, standard compliant way to advertise would be through well-known OpenID/OAuth discovery, which significantly blows up the scope. Every other mechanism would be unusual and non-standard.

> These endpoints are relative to the MCP server's base URL. For example, if the MCP server is at `https://example.com/mcp/`, the authorize endpoint would be `https://example.com/mcp/authorize`.

Reviewer: Why did you land on this instead of returning the URLs in the 401? This feels a bit more constraining for servers, while making things easier for clients—not convinced that's the right tradeoff, though I don't feel too strongly.

Author: I want to stay within HTTP verbs that libraries can easily understand. Providing URLs as part of the body is just something I've never seen and will require custom handling on the client side and also happen outside of HTTP header introspection. So I decided I rather go with what feels to me like a more HTTP-like approach.

> These endpoints are relative to the MCP server's base URL. For example, if the MCP server is at `https://example.com/mcp/`, the authorize endpoint would be `https://example.com/mcp/authorize`.

Reviewer 2: What is written here seems fine, but i am not aware that there is already a strict definition of an MCP server url structure. I don't see URL anywhere mentioned in the spec website. Is this actually defined yet?

As an example when implementing a client using the SSE client library, my impression was I would need to give a base url to hand out was directly the endpoint e.g.  `https://example.com/some-path/see` and everything in relation is undefined (e.g. the post endpoint is only discovered based on the response, or by convention, but not actually part of the spec)

Reviewer: You're correct about the current state. It may be the case that we want to make this stricter/more standardized going forward, though.

> ### Session Token Usage

Reviewer 3: > I'd like to leave room for us to decouple the auth server and the MCP server down the road.

Agreed, separation of concerns and keeping the MCP server more simpler I feel might be a better approach. https://github.com/modelcontextprotocol/specification/pull/133#discussion_r1908070902 might be a much simpler way if the server just advertises it 🤔 

>  For example, in a world where a client talks to hundreds of MCP servers, it may become valuable to have a centralized MCP OAuth server.

Agreed, that was what I was thinking as well and what I am personally looking into

> 2. replace the concept of "session token" with standard OAuth (access + refresh) tokens

> Give us the MCP server the flexibility to either forward OAuth tokens directly to the client or grant its own tokens

Loving this as well, keeps it closer to the standard and gives the flexibility / freedom for the MCP server to handle its own session by creating its own OAuth tokens (session-based auth) or pass through OAuth tokens from another provider (SSO, social login)

In the case of a spec, we could mention that an `accessToken` is always returned  and an `refreshToken` is optionally returned - this will also give the freedom for MCP servers and/or a decoupled authentication server to either pass a singular access token and optionally a way for the client to refresh the token

> ### Session Token Usage

Author: @nick-merrill This is a question that regularly comes up. The problem is that we believe the right way to do this is using the well-known OICD discovery mechanism, that adds a lot of complexity. Potentially we might defne it as '.well-known` is optional but fallbacks are predefined endpoints.

    Note over M: Generate code_verifier<br/>Calculate code_challenge<br/>Validate redirect_uri
    M->>C: Redirect to OAuth Server with code_challenge
    C->>A: OAuth Authorization Request
    A->>M: Authorization Code (via redirect)
    Note over M: Use stored code_verifier
    M->>A: Exchange Code + code_verifier
    A->>M: Access Token + Refresh Token
    M->>C: Redirect to validated redirect_uri with session token
    Note over C,M: Begin standard MCP message exchange
```

## Transport Upgrade for Authentication

For STDIO transports that require authentication, servers **MUST** use the transport upgrade mechanism during initialization to direct clients to an HTTP-based MCP server as defined in the [lifecycle specification]({{< ref "/specification/draft/basic/lifecycle" >}}). The server includes an `upgrade` field in the initialize response:

```json
{
  "jsonrpc": "2.0",
  "id": 1,
  "result": {
    "protocolVersion": "2024-11-05",
    "upgrade": {
      "endpoint": "http://new-server.example.com:8080",
      "transport": "http+sse"
    },
    "serverInfo": {
      "name": "ExampleServer",
      "version": "1.0.0"
    }
  }
}
```

## Security Considerations

> 1. Clients **MUST** securely store tokens following OAuth 2.0 best practices

Reviewer 6: This should be Servers instead of Clients, right?

* OAuth 2.0 tokens (access token + refresh token) are stored by the MCP Server.
* MCP Clients are storing the session token

> 2. Servers **SHOULD** enforce token expiration and rotation

Reviewer: ```suggestion
2. Servers **SHOULD** expire session tokens within a reasonable time
```

3. All authorization endpoints **MUST** be served over HTTPS

> 4. Clients **SHOULD** implement refresh token flows for long-lived sessions

Reviewer: ```suggestion
4. Servers **SHOULD** implement refresh token flows for long-lived sessions
```

Clients don't control this, right?

Author: I need to think more about this and would love to get input. I think this actualy might be correct but the refresh flow is not declared. The question here is, should session tokens be refreshed? You are right that SERVERS should always refresh since they hold oauth tokens. There is an additional questions if the session token for the client should also be refreshed.

Reviewer 3: In my mind there are two options - 
* Session tokens - **Servers** would need to handle session management entirely and store the access + refresh and pass over the session token as described in the spec
* Fully managed by the client - **Client** receives the access + refresh token and manages the token usage + refresh flow without needing session management on the Server

My opinion is that the second option (fully managed by the client) would be much more simpler and provides more freedom for the spec - it keeps it simple without needing to include session management in the server implementation. which I feel like is a separate concern and gives developers for Clients the freedom to do what they need with the OAuth tokens / sessions

Reviewer 2: Ultimately, at some point, the client will need to update the auth token that it passes to the MCP server in the request header.

If the client is not doing it's own refreshes, then that implies the server needs to return back the token to the client. I don't *think* that is right, otherwise why would the client pass anything at all....

I'm wondering if the client actually needs to do /all/ the steps here, and the server not do any of the exchange/refresh. I realize the motivation behind the current proposal since the common case is the server is acting as a client to a third party.

Reviewer 3: Ahhh wait, correct me if I am wrong, but is the motivation spec is for authentication for the MCP server to an external resource or is it authentication for the MCP client to MCP server 🤔?

Reading this spec more, it seems to be more of the former (MCP Server -> External Resource), since the MCP server would be managing the refresh token

Reviewer 2: I see this now, I may have missed this, and I was assuming the session token was the same thing as the oauth token.

The case I'm looking at is where there MCP Server owns the external resource, which i realize is not the only case. My assumption is that if the client owns the token lifecycle, it can use the servers auth if needed, or the server acts as a pass through. The client library already has to manage a session token, so if its just the OAuth token then there is no additional state for the server to manage (persisting beyond the lifetime of the transport connection lifetime).

I think it may be worth clarifying the principles informing the design choice. My impression is there is an assumption about trusting servers more than clients which I have not internalized yet.

Reviewer: The general problem to be solved here is:
1. MCP _clients_ that want to allow users to integrate with…
2. Arbitrary third-party services, mediated by…
3. MCP _servers_

The clients will have no foreknowledge of the MCP servers or third-party services, so it's difficult for them to (e.g.) embed a working OAuth client ID and secret for that service. The MCP server, however, would be purpose-built to support the 3P service, so can contain those details.

Most of what we're trying to figure out is how to delegate auth from client to server, while providing a secure way for clients to still utilize the MCP resources via the server.

Reviewer: To the specific question(s) above: my understanding is that session tokens don't refresh, and will eventually expire. If that happens, the client should restart the auth flow—at which point, the _server_ could utilize its OAuth refresh token.

Reviewer 2: Thank you, I now have the complete mental model.  Not sure if this is appropriate to include in the spec but it could make a nice intro on the problem being solved here or assumptions made.

Not to anchor on home assistant too much here, but i realize it is fairly similar in that regard. I wrote up a quick TL;DR of [how it works](https://gist.github.com/allenporter/c50b35f4361d6ef5176e4d4c2846b69b). This also touches on Nick's comment about using OAuth for the client/server auth too.

Reviewer 7: @jspahrsummers - your formulation of the problem statement here is quite helpful and agreed that it would be good to add it to the spec / documentation as an introduction (cc @dsp-ant)

Reviewer 3: @jspahrsummers Makes sense - agreed that clients will not have knowledge so there needs to be some sort of way of retrieving those, either through the MCP server itself **or** a separate auth service that the MCP server potentially advertises. Given the other discussions below, I'm feeling it should be a separate auth service while keeping the MCP server as a "resource" server according the the OAuth spec.

This also opens up the case where MCP servers that require third-party service resources can either have their own separate auth service or even at a simpler level, advertise the third-party's auth service to retrieve the tokens for the client to make authenticated calls with

Author: > @jspahrsummers Makes sense - agreed that clients will not have knowledge so there needs to be some sort of way of retrieving those, either through the MCP server itself **or** a separate auth service that the MCP server potentially advertises. Given the other discussions below, I'm feeling it should be a separate auth service while keeping the MCP server as a "resource" server according the the OAuth spec.

I am not sure I fully follow. Can you write up a proposal or a diagram?

One thing I want to highlight. If we add proper token exchange and refresh between client and server to this proposal, we get a nice property: An MCP server can just be it's own oauth service OR it can choose to forward to another OAuth service. Both have the same flow if we get this correct and hence it becomes easy to chain as well as simple to the client.

Reviewer 3: > I am not sure I fully follow. Can you write up a proposal or a diagram?
> 
> One thing I want to highlight. If we add proper token exchange and refresh between client and server to this proposal, we get a nice property: An MCP server can just be it's own oauth service OR it can choose to forward to another OAuth service. Both have the same flow if we get this correct and hence it becomes easy to chain as well as simple to the client.

That makes a lot of sense and actually keeps it simpler for clients to implement with - just an interface to either be it's own OAuth service or forward to another if not - especially with your comment wrt https://github.com/modelcontextprotocol/specification/pull/133#discussion_r1911107905

Related, given @nick-merrill comment https://github.com/modelcontextprotocol/specification/pull/133#discussion_r1913421367 - completely agree that this could be the spec for now to move things forward too - just that at the cost that the MCP server has to implement the auth logic in this spec as mentioned to be part of the OAuth flow, but I think we are all in alignment that this is the best approach for now

WRT a diagram - was thinking a long the lines of @jaredhanson diagrams in his comment https://github.com/modelcontextprotocol/specification/pull/133#pullrequestreview-2541159319

> 5. Servers **MUST** validate redirect URIs to prevent open redirect vulnerabilities

Reviewer: How does a server do this validation?

Reviewer 6: +1
It's not clear what's the mechanism for MCP Clients to register their redirect URIs in the MCP Server.

Reviewer 8: I also share this concern - we talked at one point about using the Origin header, however I'm fuzzy on the details on how this would work. I can imagine some potential stateful ways of solving it but haven't thought it through properly

Reviewer 3: I believe this is the whitelisting of redirect URIs in the OAuth process - in this case I believe the developers of the MCP server needs to include some sort of way for MCP Clients (ie. Claude) to whitelist specific redirect URIs that the MCP Server auth will accept during the OAuth process. This will probably be some sort of registration step before the OAuth process can begin

However, if we letting **any** MCP clients authenticate without another pre-step to register their redirect URIs, this validation isn't necessary

Perhaps we don't include this validation in the spec - but leave it for developers that implement their OAuth process to include this whitelisting if they require it 🤔

---
title: Authentication
type: docs
weight: 15
---

{{< callout type="info" >}} **Protocol Revision**: draft {{< /callout >}}

## 1. Introduction

### 1.1 Purpose and Scope

The Model Context Protocol provides authentication capabilities at the transport level,
enabling MCP clients to make requests to restricted resources on behalf of resource
owners. This specification defines the authentication flow for HTTP+SSE transport.

### 1.2 Protocol Requirements

Authentication is **OPTIONAL** for MCP implementations. Implementations using HTTP+SSE
transport **SHOULD** conform to this specification, while those using alternative
transports **MUST** follow established security best practices for their protocol.

For deployments supporting third-party clients via HTTP+SSE, implementations **SHOULD**

> expect authentication as defined here. Controlled environment deployments like corporate

Author: I think we can just remove that honestly

networks **MAY** implement alternative methods such as certificate-based authentication.

### 1.3 Standards Compliance

This authentication mechanism is based on established specifications listed below, but
implements a selected subset of their features to ensure security and interoperability
while maintaining simplicity:

- [OAuth 2.1 IETF DRAFT](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-v2-1-12)
- OAuth 2.0 Authentication Server Metadata
  ([RFC8414](https://datatracker.ietf.org/doc/html/rfc8414))
- OAuth 2.0 Dynamic Client Registration Protocol
  ([RFC7591](https://datatracker.ietf.org/doc/html/rfc7591))

## 2. Authentication Flow

### 2.1 Overview

MCP is designed to support interactive, user-facing developer tools that connect to
arbitrary servers. This leads to two key design decisions:

> 1. All MCP implementations **MUST** implement OAuth 2.1 with appropriate security
>    measures for both confidential and public clients.
> 
> 2. All MCP implementations **SHOULD** support the OAuth 2.0 Dynamic Client Registration
>    Protocol ([RFC7591](https://datatracker.ietf.org/doc/html/rfc7591)).
> 
> 3. All MCP implementations **SHOULD** implement OAuth 2.0 Authentication Server Metadata
>    ([RFC8414](https://datatracker.ietf.org/doc/html/rfc8414)). Servers that do not
>    support Authentication Server Metadata **MUST** follow the default URI schema.

Reviewer: ```suggestion
1. MCP auth implementations **MUST** implement OAuth 2.1 with appropriate security
   measures for both confidential and public clients.

2. MCP auth implementations **SHOULD** support the OAuth 2.0 Dynamic Client Registration
   Protocol ([RFC7591](https://datatracker.ietf.org/doc/html/rfc7591)).

3. MCP servers **SHOULD** and MCP clients **MUST** implement OAuth 2.0 Authorization Server Metadata
   ([RFC8414](https://datatracker.ietf.org/doc/html/rfc8414)). Servers that do not
   support Authorization Server Metadata **MUST** follow the default URI schema.
```

### 2.2 Basic OAuth 2.1 Authentication

> When authentication is required, servers **MUST** respond with HTTP 401 Unauthorized. The

Reviewer: ```suggestion
When authorization is required and has not been provided (or is not valid), servers **MUST** respond with HTTP 401 Unauthorized. The
```

basic
[OAuth 2.1 IETF DRAFT](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-v2-1-12)
authentication flow with PKCE comprises the following sequence:

```mermaid
sequenceDiagram
    participant B as User-Agent (Browser)
    participant C as Client
    participant M as MCP Server

    C->>M: MCP Request
    M->>C: HTTP 401 Unauthorized
    Note over C: Generate code_verifier and code_challenge
    C->>B: Open browser with authorization URL + code_challenge
    B->>M: GET /authorize
    Note over M: User logs in and authorizes
    M->>B: Redirect to callback URL with auth code
    B->>C: Callback with authorization code
    C->>M: Token Request with code + code_verifier

>     M->>C: Access Token + Refresh Token

Reviewer: ```suggestion
    M->>C: Access Token (+ Refresh Token)
```

    C->>M: MCP Request with Access Token
    Note over C,M: Begin standard MCP message exchange
```

### 2.3 Server Metadata Discovery

> For server capability discovery, MCP clients _SHOULD_ follow the OAuth 2.0 Authorization
> Server Metadata protocol defined in
> [RFC8414](https://datatracker.ietf.org/doc/html/rfc8414).

Reviewer: If a server supports this, a client needs to follow it—otherwise, it will try to access default endpoints and get very confused.
```suggestion
For server capability discovery, MCP clients _MUST_ follow the OAuth 2.0 Authorization
Server Metadata protocol defined in
[RFC8414](https://datatracker.ietf.org/doc/html/rfc8414).
```

> [RFC8414](https://datatracker.ietf.org/doc/html/rfc8414).

Reviewer: I suggest adding a separate paragraph which emphasizes that this is optional for servers.

        Note over C: Fall back to default endpoints
    end
    Note over C: Continue with authentication flow
```

#### 2.3.1 Fallbacks for Servers without Metadata Discovery

For servers that do not implement OAuth 2.0 Authentication Server Metadata, clients
**MUST** use the following default endpoint paths relative to the server's base URL:

| Endpoint               | Default Path | Description                          |
| ---------------------- | ------------ | ------------------------------------ |
| Authorization Endpoint | /authorize   | Used for authorization requests      |
| Token Endpoint         | /token       | Used for token exchange & refresh    |
| Registration Endpoint  | /register    | Used for dynamic client registration |

Clients **MUST** first attempt to discover endpoints via the metadata document before
falling back to default paths. When using default paths, all other protocol requirements
remain unchanged.

### 2.3 Client Registration Requirements

MCP defines two distinct registration scenarios based on the redirect URI:

#### 2.3.1 Localhost Redirect URIs

When using localhost redirect URIs (http://localhost:{port} or http://127.0.0.1:{port}),
clients:

> - Do **NOT** require dynamic registration

Reviewer: ```suggestion
- Dynamic registration is **OPTIONAL** (a client ID is not required)
```

> - **MUST** use PKCE

Reviewer: I think we require this everywhere, so it's redundant here.

- **MAY** proceed directly to authorization
- **MUST NOT** require client secrets

This exception for localhost is explicitly supported by OAuth 2.1 for public clients and
provides a secure flow through the combination of PKCE and localhost-only redirects.

#### 2.3.2 Non-Localhost Redirect URIs

For all other redirect URIs, dynamic client registration is **REQUIRED**. This provides a
standardized way for clients to automatically register with new servers, which is crucial
for MCP because:

- Clients cannot know all possible servers in advance
- Manual registration would create friction for users
- It enables seamless connection to new servers
- Servers can implement their own registration policies

#### 2.3.3 Registration Process

For non-localhost scenarios, clients **SHOULD** implement the OAuth 2.0 Dynamic Client
Registration Protocol as defined in
[RFC7591](https://datatracker.ietf.org/doc/html/rfc7591). This enables automatic
registration with the server's registration endpoint to obtain the necessary client
credentials.

The registration request and response formats, along with all requirements and error

> handling, are defined in RFC7591.

Reviewer: This section feels redundant (and with the "SHOULD," contradictory). If this is just meant to point to the RFC, can we link it in the above sections?

### 2.4 Authentication Flow Steps

The complete authentication flow proceeds as follows:

> 1. Client discovers server metadata (if supported)

Reviewer: ```suggestion
1. Client discovers server metadata (if available)
```

> 2. For non-localhost redirect URIs, client **MUST** register dynamically with server

Reviewer: As above, I think this should be recommended but not required. Client ID and client secret could be used here.

6. Server returns authorization code
7. Client exchanges code and code_verifier for tokens
8. Client uses access token for API requests

**NOTE**: Dynamic registration (step 2) is only required when using non-localhost
redirect URIs. Clients using localhost redirect URIs may skip this step.

Flow diagram:

```mermaid
sequenceDiagram
    participant B as User-Agent (Browser)
    participant C as Client
    participant M as MCP Server

    C->>M: GET /.well-known/oauth-authorization-server
    alt Server Supports Discovery
        M->>C: Authorization Server Metadata
    else No Discovery
        M->>C: 404 (Use default endpoints)
    end

    alt Non-Localhost Redirect URI
        C->>M: POST /register
        M->>C: Client Credentials
    end

    Note over C: Generate PKCE Parameters
    C->>B: Open browser with authorization URL + code_challenge
    B->>M: Authorization Request
    Note over M: User authenticates/authorizes
    M->>B: Redirect to callback with authorization code
    B->>C: Authorization code callback
    C->>M: Token Request + code_verifier
    M->>C: Access Token + Refresh Token
    C->>M: API Requests with Access Token
```

#### 2.4.1 Decision Flow Overview

> The authentication flow follows this decision tree:

Author: I did and it looks awful :p. I dont like it, but i can do that

1. Server Metadata Discovery

   1. Check if server supports metadata discovery at
      /.well-known/oauth-authorization-server
      - If available: Use endpoints from metadata document
      - If not available: Use default endpoint paths
        - /authorize for authorization
        - /token for token exchange
        - /register for client registration

2. Client Registration Check

   1. Check if using localhost redirect URI
      - If localhost: Skip registration, proceed to OAuth flow
      - If non-localhost: Registration required
        1. Check for registration endpoint
           - If available: Perform dynamic registration

>            - If not available: Manual registration required

Reviewer: Is "manual registration" a well-understood thing?

3. OAuth 2.1 Flow
   1. Generate PKCE code_verifier and code_challenge
   2. Request authorization with code_challenge
   3. User authenticates and authorizes
   4. Exchange authorization code for tokens
   5. Use access token for API requests

### 2.5 Access Token Usage

#### 2.5.1 Token Requirements

Access token handling **MUST** conform to
[OAuth 2.1 Section 5](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-v2-1-12#section-5)
requirements for resource requests. Specifically:

1. Access tokens **MUST** be included using one of the two methods defined in

>    [Section 5.1](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-v2-1-12#section-5.1)

Reviewer: I thought we agreed to only support the header?

response with a WWW-Authenticate header indicating the error.

### 2.6 Security Considerations

The following security requirements **MUST** be implemented:

1. Clients **MUST** securely store tokens following OAuth 2.0 best practices
2. Servers **SHOULD** enforce token expiration and rotation
3. All authorization endpoints **MUST** be served over HTTPS
4. Servers **MUST** validate redirect URIs to prevent open redirect vulnerabilities
5. Redirect URIs **MUST** be either localhost URLs or HTTPS URLs

### 2.7 Error Handling

Servers **MUST** return appropriate HTTP status codes for authentication errors:

| Status Code | Description  | Usage                                      |
| ----------- | ------------ | ------------------------------------------ |
| 401         | Unauthorized | Authentication required or token invalid   |
| 403         | Forbidden    | Invalid scopes or insufficient permissions |
| 400         | Bad Request  | Malformed authentication request           |

### 2.8 Implementation Requirements

1. Implementations **MUST** follow OAuth 2.1 security best practices
2. PKCE is **REQUIRED** for all clients

> 3. Refresh token rotation **SHOULD** be implemented for enhanced security

Reviewer: ```suggestion
3. Token rotation **SHOULD** be implemented for enhanced security
```

    B->>T: Authorization Request
    Note over T: User authenticates
    T->>B: Redirect to MCP Server callback
    B->>M: Authorization code
    M->>T: Exchange code for token
    T->>M: Third-party access token
    Note over M: Generate bound MCP token
    M->>B: Redirect to MCP Client callback
    B->>C: MCP authorization code
    C->>M: Exchange code for token
    M->>C: MCP access token
```

#### 2.9.3 Session Binding Requirements

MCP servers implementing third-party authentication **MUST**:

1. Maintain secure mapping between third-party tokens and issued MCP tokens
2. Validate third-party token status before honoring MCP tokens
3. Implement appropriate token lifecycle management
4. Handle third-party token expiration and renewal

#### 2.9.4 Security Considerations

When implementing third-party authentication, servers **MUST**:

1. Validate all redirect URIs
2. Securely store third-party credentials
3. Implement appropriate session timeout handling
4. Consider security implications of token chaining
5. Implement proper error handling for third-party auth failures

## 3. Best Practices

#### 3.1 Local clients as Public OAuth 2.1 Clients

> We strongly recommend that local clients implement OAuth 2.1 as public client:

Reviewer: Just curious, why is this preferable to a local client being a confidential client?

Author: Because a local client cannot store secrets securely. This is equivalent to a Oauth 2.0 native application which are treated as public clients.

1. Utilizing code challenges (PKCE) for authorization requests to prevent interception
   attacks
2. Implementing secure token storage appropriate for the local system
3. Following token refresh best practices to maintain sessions
4. Properly handling token expiration and renewal

#### 3.2 Authentication Metadata Discovery

We strongly recommend, that all clients implement metadata discovery. This reduces the

> need for users to provide endpoints either manually or fallback to the defined defaults.

Reviewer: I think we should delete this section—it's covered above.

#### 3.3 Dynamic Client Registration

Since clients do not know the set of MCP servers in advance, we strongly recommend the
implementation of dynamic client registration. This allows applications to automatically

> register with the MCP server, and removing the need for users to obtain client ids

Reviewer: ```suggestion
register with the MCP server, and removes the need for users to obtain client IDs
```

Author: I am okay with that duplication, since its within a best practices. I think a lot of people will go back to a best practices section without reading the full spec again.

---
title: Authentication
type: docs
weight: 15
---

{{< callout type="info" >}} **Protocol Revision**: draft {{< /callout >}}

## 1. Introduction

### 1.1 Purpose and Scope

The Model Context Protocol provides authentication capabilities at the transport level,
enabling MCP clients to make requests to restricted resources on behalf of resource
owners. This specification defines the authentication flow for HTTP+SSE transport.

### 1.2 Protocol Requirements

MCP clients and servers supporting HTTP+SSE transport **SHOULD** implement this

> specification, though they **MAY** support additional authentication mechanisms.

Reviewer: Is this a "should" even if the server doesn't want/need authentication? Or is the idea that basically all remote servers supporting third-party clients probably need to support auth anyway (e.g., to prevent abuse)?

Author: I think the idea is, if you want a local only one you can skip it.

Implementations using alternative transports **MAY** implement authentication according
to the best practices for their specific transport.

For deployments supporting third-party clients or servers via HTTP+SSE, implementations
**SHOULD** expect authentication as defined in this document. However, controlled
environment deployments like corporate networks **MAY** implement alternative methods
such as certificate-based authentication.

### 1.3 Standards Compliance

This authentication mechanism implements following specifications but recommends a

> specific subsets:

Reviewer: I'm having trouble parsing this phrase…

- [OAuth 2.1 IETF DRAFT](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-v2-1-12)
- OAuth 2.0 Authentication Server Metadata
  ([RFC8414](https://datatracker.ietf.org/doc/html/rfc8414))
- OAuth 2.0 Dynamic Client Registration Protocol
  ([RFC7591](https://datatracker.ietf.org/doc/html/rfc7591))

## 2. Authentication Flow

### 2.1 Overview

MCP is designed to support interactive, user-facing developer tools that connect to
arbitrary servers. This leads to two key design decisions:

1. All MCP clients **MUST** be treated as public OAuth 2.1 clients, as they cannot

>    securely store client secrets when distributed to end users

Reviewer: This doesn't make sense to me. A backend API could be an MCP client.

Author: Okay thats sensible.

2. PKCE **MUST** be implemented by all clients to prevent authorization code interception

>    attacks, which is especially important for locally-running tools

Reviewer: Refer to OAuth 2.1 here? (Notably it doesn't call it "PKCE" anymore, just incorporates it directly.)

3. All MCP implementations **MUST** support the OAuth 2.0 Dynamic Client Registration
   Protocol ([RFC7591](https://datatracker.ietf.org/doc/html/rfc7591)), with the
   exception of servers that only listen on loopback interfaces (e.g. localhost for

>    native application authentication).

Reviewer: I thought we were only going to _recommend_ this. Why is it required here?

Author: We agreed.

```mermaid
sequenceDiagram
    participant B as User-Agent (Browser)
    participant C as Client
    participant M as MCP Server

    C->>M: MCP Request
    M->>C: HTTP 401 Unauthorized
    Note over C: Generate code_verifier and code_challenge
    C->>B: Open browser with authorization URL + code_challenge
    B->>M: GET /authorize
    Note over M: User logs in and authorizes
    M->>B: Redirect to callback URL with auth code
    B->>C: Callback with authorization code
    C->>M: Token Request with code + code_verifier
    M->>C: Access Token + Refresh Token
    C->>M: MCP Request with Access Token
    Note over C,M: Begin standard MCP message exchange
```

### 2.3 Server Metadata Discovery

Since MCP clients need to connect to previously unknown servers, automated server
capability discovery is essential. The metadata discovery mechanism allows clients to:

- Locate the OAuth endpoints without manual configuration
- Determine supported authentication features
- Adapt to server-specific requirements
- Enable zero-configuration connections to new servers

The discovery flow is illustrated below:

```mermaid
sequenceDiagram
    participant C as Client
    participant S as Server

>     C->>S: GET /.well-known/oauth-authorization-server

Author: We should keep it in accordance to the Server Metadata spec, which defines the `/.well-known/oauth-authorization-server` url. If we need something MCP specific and go and register a well known for MCP. 

Since we deal with well known for oauth, and keep oauth separate from the MCP protocol, we should not requiring sending that header and just treat it as normal oauth.

>     C->>S: GET /.well-known/oauth-authorization-server

Reviewer: I'm fine keeping this well-known URL, but I feel strongly that we should include a recommendation to send that header. This would allow servers to issue different versions of this file for MCP vs. REST, for example.

Since it's only a recommendation, clients do not have to include it if it's technically difficult.

>     C->>S: GET /.well-known/oauth-authorization-server

Author: I'll add taht header to the Spec as a recommendation.

remain unchanged.

Example default endpoints for https://mcp.example.com:

```
https://mcp.example.com/authorize
https://mcp.example.com/token
https://mcp.example.com/register
```

### 2.3 Client Registration Requirements

MCP defines two distinct registration scenarios based on the redirect URI:

#### 2.3.1 Localhost Redirect URIs

When using localhost redirect URIs (http://localhost:{port} or http://127.0.0.1:{port}),
clients:

- Do **NOT** require dynamic registration
- **MUST** use PKCE
- **MAY** proceed directly to authorization
- **MUST NOT** require client secrets

This exception for localhost is explicitly supported by OAuth 2.1 for public clients and
provides a secure flow through the combination of PKCE and localhost-only redirects.

#### 2.3.2 Non-Localhost Redirect URIs

For all other redirect URIs, dynamic client registration is **REQUIRED**. This provides a
standardized way for clients to automatically register with new servers, which is crucial
for MCP because:

- Clients cannot know all possible servers in advance
- Manual registration would create friction for users
- It enables seamless connection to new servers

> - Servers can implement their own registration policies

Reviewer: @dsp-ant ☝️ I still think this

    end

    Note over C: Generate PKCE Parameters
    C->>B: Open browser with authorization URL + code_challenge
    B->>M: Authorization Request
    Note over M: User authenticates/authorizes
    M->>B: Redirect to callback with authorization code
    B->>C: Authorization code callback
    C->>M: Token Request + code_verifier
    M->>C: Access Token + Refresh Token
    C->>M: API Requests with Access Token
```

### 2.5 Access Token Usage

#### 2.5.1 Token Requirements

Access token handling **MUST** conform to
[OAuth 2.1 Section 5](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-v2-1-12#section-5)
requirements for resource requests. Specifically:

1. Access tokens **MUST** be included using one of the two methods defined in
   [Section 5.1](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-v2-1-12#section-5.1)

2. The preferred method is using the Authorization header with Bearer scheme per
   [Section 5.1.1](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-v2-1-12#section-5.1.1):

```
Authorization: Bearer <access-token>
```

3. Form-encoded body parameters per
   [Section 5.1.2](https://datatracker.ietf.org/doc/html/draft-ietf-oauth-v2-1-12#section-5.1.2)

>    **MAY** be supported as a fallback

Reviewer: Is there any reason to not just require the `Authorization` header? I don't think we have a reason to support the form-encoded content variant.

Author: Good point.
